File: python/binary_search.py


def bs(arr, x, lower_bound=True):
    '''
    Arguments:
        arr: A sorted array of elements
        x: The element to search for
        lower_bound: If set to True, finds the first element >= x, else finds the first element > x
    Returns:
        The index of the found element, or the length of the array if not found
    '''

    hi = len(arr)
    lo = 0
    mid = (hi + lo) // 2

    while hi > lo:
        if (lower_bound and arr[mid] >= x) or (not lower_bound and arr[mid] > x):
            hi = mid
        else:
            lo = mid + 1
        mid = (hi + lo) // 2
    
    return mid


def main():

    return 0

if __name__ == "__main__":
    main()

File: rust_algorithms/src/binary_search.rs

// Returns the index of the first element not less than x
fn bs<T: PartialOrd>(arr: &Vec<T>, x: T) -> usize {

    let mut hi: usize = arr.len();
    let mut lo: usize = 0;

    let mut mid = (hi + lo) / 2;

    while hi > lo {

        if arr[mid] >= x {
            hi = mid;
        } else {
            lo = mid + 1;
        }
        mid = (hi + lo) / 2;

    }
    mid
}

